### PR TITLE
crimson/net: Complete the refactor to std::unique_ptr inside Messenger

### PR DESCRIPTION
**Pull request by AmnonHanuhov**
This PR should be the last of the 3 PRs that refactor Messenger to use `std::unique_ptr`. It includes a few more changes to the shared code in `PeeringState` as well as refactoring any last users of the previous version of `make_message` and wrapper functions to `Connection::send()`

Signed-off-by: Amnon Hanuhov <ahanukov@redhat.com>

### CODE REVIEW

File: src/osd/PeeringState.h [hunk 1]
```diff
@@ -61,7 +61,11 @@ struct PeeringCtx;
 
 // [primary only] content recovery state
 struct BufferedRecoveryMessages {
+  #if defined(WITH_SEASTAR)
```
Comment (rzarzynski) [inline src/osd/PeeringState.h:64]: nit: I believe our convention recommends to put such macros at the 0th indent level.

File: src/crimson/net/Connection.h [hunk 2]
```diff
@@ -127,8 +127,6 @@ class Connection : public seastar::enable_shared_from_this<Connection> {
 
   /// send a message over a connection that has completed its handshake
   virtual seastar::future<> send(MessageURef msg) = 0;
-  // The version with MessageRef will be dropped in the future
-  virtual seastar::future<> send(MessageRef msg) = 0;
```
Comment (rzarzynski) [inline src/crimson/net/Connection.h:131]: :+1:

**Review by rzarzynski — COMMENTED** [review]
Overall looks good!

File: src/crimson/net/Protocol.cc [hunk 3]
```diff
@@ -204,7 +204,7 @@ void Protocol::ack_writes(seq_num_t seq)
   }
   while (!conn.sent.empty() && conn.sent.front()->get_seq() <= seq) {
     logger().trace("{} got ack seq {} >= {}, pop {}",
-                   conn, seq, conn.sent.front()->get_seq(), conn.sent.front());
+                   conn, seq, conn.sent.front()->get_seq(), std::move(*conn.sent.front()));
```
Comment (cyx1231st) [inline src/crimson/net/Protocol.cc:207]: Why use move here?

**Review by cyx1231st — COMMENTED** [review]
Need to fix the unit-test `src/test/crimson/test_messenger.cc` and make sure it is passing.

File: src/osd/PeeringState.h (hunk 1 above)
Comment (AmnonHanuhov) [reply src/osd/PeeringState.h]: Thanks for the heads up, fixed it.

File: src/crimson/osd/osd_operations/client_request.cc [hunk 4]
```diff
@@ -139,9 +139,9 @@ ClientRequest::interruptible_future<>
 ClientRequest::process_pg_op(
   Ref<PG> &pg)
 {
-  return pg->do_pg_ops(m)
-    .then_interruptible([this, pg=std::move(pg)](Ref<MOSDOpReply> reply) {
-      return conn->send(reply);
+  return std::move(pg->do_pg_ops(m))
```
[the remark this answers is not in the document]
Comment (AmnonHanuhov) [reply src/crimson/osd/osd_operations/client_request.cc]: No reason, fixed it.

File: src/crimson/net/Protocol.cc (hunk 3 above)
Comment (AmnonHanuhov) [reply src/crimson/net/Protocol.cc]: Got rid of it, thanks.

**Comment by AmnonHanuhov** [comment]
Fixed the unit-test `src/test/crimson/test_messenger.cc` and merged #41916 @cyx1231st

**Review by cyx1231st — APPROVED** [review]

**Review by tchaikov — APPROVED** [review]

File: src/crimson/net/Connection.h (hunk 2 above)
Comment (tchaikov) [reply src/crimson/net/Connection.h:131]: this breaks the build of `test-seastar-echo`.

File: src/crimson/osd/shard_services.cc [hunk 5]
```diff
@@ -85,9 +85,8 @@ void ShardServices::handle_conf_change(const ConfigProxy& conf,
   }
 }
 
-template<class MsgT>
-seastar::future<> ShardServices::do_send_to_osd(
-  int peer, MsgT m, epoch_t from_epoch)
+seastar::future<> ShardServices::send_to_osd(
+  int peer, MessageURef m, epoch_t from_epoch)
```
Comment (tchaikov) [inline src/crimson/osd/shard_services.cc:89]: could you please remove trailing space?

**Review by tchaikov — CHANGES_REQUESTED** [review]
could you please address the FTBFS? see https://shaman.ceph.com/builds/ceph/wip-kefu-testing-2021-06-19-1642/d1ef7e90506d79b7bf7cff7ea131f4ab7a2fb740/crimson/274164/

**Comment by AmnonHanuhov** [comment]
> could you please address the FTBFS? see https://shaman.ceph.com/builds/ceph/wip-kefu-testing-2021-06-19-1642/d1ef7e90506d79b7bf7cff7ea131f4ab7a2fb740/crimson/274164/

On it. Looks like there are still some crimson tests that use `ceph::make_message()` instead of `crimson::make_message()`, will release a PR fixing that.

**Comment by tchaikov** [comment]
> > could you please address the FTBFS? see https://shaman.ceph.com/builds/ceph/wip-kefu-testing-2021-06-19-1642/d1ef7e90506d79b7bf7cff7ea131f4ab7a2fb740/crimson/274164/
> 
> On it. Looks like there are still some crimson tests that use `ceph::make_message()` instead of `crimson::make_message()`, will release a PR fixing that.

https://github.com/ceph/ceph/pull/41861#discussion_r654786567

**Comment by AmnonHanuhov** [comment]
> > > could you please address the FTBFS? see https://shaman.ceph.com/builds/ceph/wip-kefu-testing-2021-06-19-1642/d1ef7e90506d79b7bf7cff7ea131f4ab7a2fb740/crimson/274164/
> > 
> > 
> > On it. Looks like there are still some crimson tests that use `ceph::make_message()` instead of `crimson::make_message()`, will release a PR fixing that.
> 
> [#41861 (comment)](https://github.com/ceph/ceph/pull/41861#discussion_r654786567)

See #41942

File: src/crimson/osd/shard_services.cc (hunk 5 above)
Comment (AmnonHanuhov) [reply src/crimson/osd/shard_services.cc]: Fixed and repushed.

**Review by tchaikov — APPROVED** [review]

**Review by tchaikov — CHANGES_REQUESTED** [review]
the comment was not addressed in this PR.

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-06-18_12:04:35-rados-wip-kefu-testing-2021-06-18-1414-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-06-19_08:35:14-rados-wip-kefu-testing-2021-06-18-1414-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-06-19_11:27:01-rados-wip-kefu-testing-2021-06-19-1642-distro-basic-smithi/

i don't think the failures were related to this change.

as i reran the failed tests without this change:

- https://pulpito.ceph.com/kchai-2021-06-19_12:18:10-rados-wip-kefu-testing2-2021-06-19-1707-distro-basic-smithi/

**Comment by AmnonHanuhov** [comment]
> the comment was not addressed in this PR.

Added two commits to this PR fixing the issue.

**Review by rzarzynski — APPROVED** [review]

**Review by tchaikov — APPROVED** [review]

**Pull request merged**